### PR TITLE
add library search mz filter(>300)

### PR DESCRIPTION
**Pull request by Yuling1996**

### CODE REVIEW

File: MassSpectrometry/MzSpectra/SpectralSimilarity.cs [hunk 1]
```diff
@@ -161,8 +176,12 @@ private double[] NormalizeSpectrumSum(double[] spectrum)
         #region similarityMethods
 
         //The cosine similarity returns values between 1 and -1 with 1 being closes and -1 being opposite and 0 being orthoganal
-        public double CosineSimilarity()
+        public double? CosineSimilarity()
```
Comment (trishorts) [inline MassSpectrometry/MzSpectra/SpectralSimilarity.cs:179]: do we really want null?

**Comment by trishorts** [comment]
we need a series of unit tests that return values when a cut-off is used. it should be different than when no cutoff is used. the number of pairs should change when there are mz values below cuttoff. if you are going to return null, then we have to check that null value is returned for each instance in a unit test.

File: MassSpectrometry/MzSpectra/SpectralSimilarity.cs (hunk 1 above)
Comment (Yuling1996) [reply MassSpectrometry/MzSpectra/SpectralSimilarity.cs:179]: we don't want null but it happens when all mz are less than 300 in theoretical(library) peaks

Comment (Yuling1996) [reply MassSpectrometry/MzSpectra/SpectralSimilarity.cs:179]: I mean it happens that there is no intensity pair when all mz are less than 300 in theoretical(library) peaks. I need to assign a value to it

**Comment by Yuling1996** [comment]
> we need a series of unit tests that return values when a cut-off is used. it should be different than when no cutoff is used. the number of pairs should change when there are mz values below cuttoff. if you are going to return null, then we have to check that null value is returned for each instance in a unit test.

For null test, does line 308-318 in Test/TestSpectralSimilarity.cs make sense?

**Comment by Yuling1996** [comment]
> we need a series of unit tests that return values when a cut-off is used. it should be different than when no cutoff is used. the number of pairs should change when there are mz values below cuttoff. if you are going to return null, then we have to check that null value is returned for each instance in a unit test.

There is new test TestAllSpectrumSimilaritiesWithDefaultedMzFilter() with cut off used. Did I miss something?

**Comment by trishorts** [comment]
> > we need a series of unit tests that return values when a cut-off is used. it should be different than when no cutoff is used. the number of pairs should change when there are mz values below cuttoff. if you are going to return null, then we have to check that null value is returned for each instance in a unit test.
> 
> For null test, does line 308-318 in Test/TestSpectralSimilarity.cs make sense?

im not sure. definitely line 300 intensity pairs is zero, not null. After that seems maybe okay. But, the secondary has mz values all above 300, so i guess it depends if we use secondary array to make intensity pairs or not. The

**Comment by trishorts** [comment]
> > we need a series of unit tests that return values when a cut-off is used. it should be different than when no cutoff is used. the number of pairs should change when there are mz values below cuttoff. if you are going to return null, then we have to check that null value is returned for each instance in a unit test.
> 
> There is new test TestAllSpectrumSimilaritiesWithDefaultedMzFilter() with cut off used. Did I miss something?

is default 0 or 300? if 300, then that is correct. i didn't see that.

**Comment by Yuling1996** [comment]
> > > we need a series of unit tests that return values when a cut-off is used. it should be different than when no cutoff is used. the number of pairs should change when there are mz values below cuttoff. if you are going to return null, then we have to check that null value is returned for each instance in a unit test.
> > 
> > 
> > There is new test TestAllSpectrumSimilaritiesWithDefaultedMzFilter() with cut off used. Did I miss something?
> 
> is default 0 or 300? if 300, then that is correct. i didn't see that.

default is 300

**Comment by Yuling1996** [comment]
> > > we need a series of unit tests that return values when a cut-off is used. it should be different than when no cutoff is used. the number of pairs should change when there are mz values below cuttoff. if you are going to return null, then we have to check that null value is returned for each instance in a unit test.
> > 
> > 
> > For null test, does line 308-318 in Test/TestSpectralSimilarity.cs make sense?
> 
> im not sure. definitely line 300 intensity pairs is zero, not null. After that seems maybe okay. But, the secondary has mz values all above 300, so i guess it depends if we use secondary array to make intensity pairs or not. The

what do you think that I should add?

File: MassSpectrometry/MzSpectra/SpectralSimilarity.cs [hunk 2]
```diff
@@ -105,7 +117,11 @@ private List<(double, double)> IntensityPairs(bool allPeaks)
                     }
                     index++;
                 }
-                if(index > 0)
+                if (experimental.Count == 0)
```
Comment (trishorts) [inline MassSpectrometry/MzSpectra/SpectralSimilarity.cs:120]: please explain this change

**Review by trishorts — APPROVED** [review]

**Pull request merged**